packages/crypto/src/secp256r1.rs

```diff
@@ -0,0 +1,277 @@
+
+impl From<InvalidSecp256r1SignatureFormat> for CryptoError {
+    fn from(_original: InvalidSecp256r1SignatureFormat) -> Self {
+        CryptoError::invalid_signature_format()
+    }
+}
+
+fn read_signature(data: &[u8]) -> Result<[u8; 64], InvalidSecp256r1SignatureFormat> {
+    data.try_into().map_err(|_| InvalidSecp256r1SignatureFormat)
+}
+
+/// Error raised when public key is not in one of the two supported formats:
+/// 1. Uncompressed: 65 bytes starting with 0x04
+/// 2. Compressed: 33 bytes starting with 0x02 or 0x03
+struct InvalidSecp256r1PubkeyFormat;
+
+impl From<InvalidSecp256r1PubkeyFormat> for CryptoError {
+    fn from(_original: InvalidSecp256r1PubkeyFormat) -> Self {
+        CryptoError::invalid_pubkey_format()
+    }
+}
+
+fn check_pubkey(data: &[u8]) -> Result<(), InvalidSecp256r1PubkeyFormat> {
+    let ok = match data.first() {
+        Some(0x02) | Some(0x03) => data.len() == ECDSA_COMPRESSED_PUBKEY_LEN,
+        Some(0x04) => data.len() == ECDSA_UNCOMPRESSED_PUBKEY_LEN,
+        _ => false,
+    };
+    if ok {
+        Ok(())
+    } else {
+        Err(InvalidSecp256r1PubkeyFormat)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // use elliptic_curve::rand_core::OsRng;
+    // use elliptic_curve::sec1::ToEncodedPoint;
+
+    use hex_literal::hex;
+    use p256::{
+        ecdsa::signature::DigestSigner, // trait
+        ecdsa::SigningKey,              // type alias
+        elliptic_curve::rand_core::OsRng,
+    };
+    use serde::Deserialize;
+    use sha2::Sha256;
+    use std::fs::File;
+    use std::io::BufReader;
+
+    // For generic signature verification
+    const MSG: &str = "Hello World!";
+
+    // Cosmos secp256r1 signature verification
+    // tendermint/PubKeySecp256r1 pubkey
+    const COSMOS_SECP256R1_PUBKEY_HEX: &str =
+        "049a2c7b27b132246e170dfb9167db5c5bd302033dbece2bc3f2541a6cd11851821a775f1fc6c4f89e0d019888057f0d574f1c4eb1f90a7a41c4ea9b99b538d932";
+
+    const COSMOS_SECP256R1_MSG_HEX1: &str = "6265206b696e64";
+    const COSMOS_SECP256R1_MSG_HEX2: &str = "6265206b696e64";
+    const COSMOS_SECP256R1_MSG_HEX3: &str = "6265206b696e64";
+
+    const COSMOS_SECP256R1_SIGNATURE_HEX1: &str = "453020029250fb9eb22b21b881319a123244e463a329356b75ce804fc2dda174e715104621028d009abee7d523894b425d974bc38cfae5d05cdf5a550c8eceae1f20f0c9913f0038";
+    const COSMOS_SECP256R1_SIGNATURE_HEX2: &str = "30450220658fc9271b09bd53edf3a5bd31b7bd99bd3c3de7859cd8dd1133e76ed44fcb580221009e43d091911de0fc90d22960517211f5cf6c624b326759e219326f3af807ac31";
+    const COSMOS_SECP256R1_SIGNATURE_HEX3: &str = "30450220658fc9271b09bd53edf3a5bd31b7bd99bd3c3de7859cd8dd1133e76ed44fcb580221009e43d091911de0fc90d22960517211f5cf6c624b326759e219326f3af807ac31";
+
+    // Test data originally from https://github.com/cosmos/cosmjs/blob/v0.24.0-alpha.22/packages/crypto/src/secp256k1.spec.ts#L195-L394
+    const COSMOS_SECP256R1_TESTS_JSON: &str = "./testdata/secp256r1_tests.json";
```

Comment:
Test data needs to be changed accordingly, but i don't know how to do that. Looked into [secp256r1 test in cosmos sdk](https://github.com/cosmos/cosmos-sdk/tree/main/crypto/keys/secp256r1), but i didn't find any hardcoded test case.

Reply:
Just seen this. I remember I scoured the web at the time to find good test data for secp256k1, and ended up using this ref. Maybe you can do the same.

Here, this looks promising: https://github.com/scogliani/ecc-test-vectors